### ISSUE TITLE
Support pasts as inputs in gpt2 beam search operator

### ISSUE THREAD

**Issue by HaoboGu** [microsoft/onnxruntime#12047]
Hello, I noticed that ort would support beam search operator for gpt2 model. 

I'm wondering whether this operator support pasts as inputs? In many cases, the pasts can be reused for multiple beam search runs making inference much more faster.

Here is the export script, it seems that it doesn't have `pasts` as the node input:

https://github.com/microsoft/onnxruntime/blob/ecca6f4d164b08e1e1b12aa1cd6f8c7557b0654b/onnxruntime/python/tools/transformers/convert_beam_search.py#L786-L804

**Comment by wangyems** [comment]
Hi @HaoboGu, the cached pasts are supported in beam search op - but not in the form of main graph's inputs. This op takes encoder/decoder as subgraphs in node attributes. You can take a look at the definition of the decoder: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc#L25-L31

**Comment by HaoboGu** [comment]
@wangyems So if I wanna use cached pasts, I have to write my own `export_beam_search.py`, and then add the cached pasts to graph input, am I right?

**Comment by wangyems** [comment]
> @wangyems So if I wanna use cached pasts, I have to write my own `export_beam_search.py`, and then add the cached pasts to graph input, am I right?

The current `export_beam_search.py` supports pasts. By running this script you'll get an onnx model with decoder that already containing the past as input.
If you have a customized gpt2 model, you only need to make sure that the exported decoder onnx has the pasts as inputs. see https://github.com/microsoft/onnxruntime/blob/57ac3d0a612e6117b094fb7ead63658c6f61a3d4/onnxruntime/python/tools/transformers/models/gpt2/gpt2_helper.py#L89

**Comment by tianleiwu** [comment]
@[HaoboGu](https://github.com/HaoboGu), past is used in gpt2 beam search. You can observe it when opening the decoder subgraph with Netron tool. The expected inputs for GPT-2 subgraph is like the following:
https://github.com/microsoft/onnxruntime/blob/df712d80ca33b56ce7d63ca3acdb6a4888638c58/onnxruntime/python/tools/transformers/convert_beam_search.py#L437

**Comment by HaoboGu** [comment]
@tianleiwu yeah, I know the empty past is auto-generated before doing beam search, my exported model has only those inputs:
```python
     "input_ids", 
     "max_length", 
     "min_length", 
     "num_beams", 
     "num_return_sequences", 
     "length_penalty", 
     "repetition_penalty", 
```

What I want is export past as also an option input, so that users can do beam search based on cached pasts. (In this case, `input_ids` should be ids after the given past state)

**Comment by tianleiwu** [comment]
@viboga, could you share @HaoboGu the custom operator of beam search.  I think it only need some minor changes for this scenario.

**Comment by viboga** [comment]
@HaoboGu Currently we are keeping beam search as the most commonly used simple variant. Is there any link you are referring to for your project, is past_key_values support based on a new publication?

I am working on beam search as custom OP and planning to publish it end of month which can be leveraged for customizations.

**Comment by HaoboGu** [comment]
@viboga Yeah, our project is based on transformers 4.18.0, which supports past_key_values.

---- some other suggestions below

Besides pasts, I think a multi-step prefix_vocab_mask is also useful, especially when we want to do generation **case-insensitively**.

For example, if current incomplete token is `beamsear`, it might be tokenized into `beam` and `sear`. In **case-insensitive** prediction, what we want might be `BeamSearch`. So a two-step mask of `Beam`/`beam` and `Sear*`/`sear*` is needed.

**Comment by HaoboGu** [comment]
@viboga any updates on beam search custom OP?
